src/main/java/org/intellij/sdk/codesync/codeSyncSetup/S3FileUploader.java

```diff
@@ -0,0 +1,133 @@
+package org.intellij.sdk.codesync.codeSyncSetup;
+
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import org.intellij.sdk.codesync.CodeSyncLogger;
+import org.intellij.sdk.codesync.clients.CodeSyncClient;
+import org.intellij.sdk.codesync.exceptions.FileLockedError;
+import org.intellij.sdk.codesync.exceptions.InvalidYmlFileError;
+import org.intellij.sdk.codesync.exceptions.RequestError;
+import org.intellij.sdk.codesync.files.S3UploadQueueFile;
+import org.intellij.sdk.codesync.repoManagers.OriginalsRepoManager;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class S3FileUploader {
+    S3UploadQueueFile s3UploadQueueFile;
+    Integer runCount;
+    Map<String, Object> failedFilePathsAndURLs = new HashMap<>();
+
+    final Integer maxRunCount = 10;
+
+    public S3FileUploader (S3UploadQueueFile s3UploadQueueFile) {
+        this.s3UploadQueueFile = s3UploadQueueFile;
+        this.runCount = this.s3UploadQueueFile.getRunCount();
+    }
+    public S3FileUploader (String repoPath, String branch, Map<String, Object> filePathAndURLs) {
+        this.runCount = 0;
+        this.s3UploadQueueFile = new S3UploadQueueFile(repoPath, branch, filePathAndURLs, runCount);
+    }
+
+    public void saveURLs() throws InvalidYmlFileError, FileNotFoundException {
+        try {
+            this.s3UploadQueueFile.writeYml();
+        } catch (FileLockedError error) {
+            // This would never happen, but adding it for completeness.
+            CodeSyncLogger.error("File locked error while creating a new file for S3 file upload queue.");
+        }
+    }
+
+    /*
+    Pre-Process and validate yml data. return `true` of we are good to proceed, `false` otherwise.
+    */
+    public boolean isValid() {
+        if (!this.s3UploadQueueFile.hasFiles()){
+            return false;
+        }
+
+        // If previous 3 attempts have resulted in failures then do not try again.
+        return this.runCount <= this.maxRunCount;
+    }
+
+    public void uploadToS3(String repoPath, String branchName, Map<String, Object> fileUrls) {
+        OriginalsRepoManager originalsRepoManager = new OriginalsRepoManager(repoPath, branchName);
+        CodeSyncClient codeSyncClient = new CodeSyncClient();
+
+        for (Map.Entry<String, Object> fileUrl : fileUrls.entrySet()) {
```

Comment:
I did not do that because I wanted to avoid multiple parallel processes, we already have quite a few parallel processes in the plugin and adding a few more may slow down the plugin.